### PR TITLE
NAS-119828 / 23.10 / Fix multiple crashes on attempt to unlock a dataset that has VMs on it

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

File: src/middlewared/middlewared/plugins/pool_/unlock.py [hunk 1]
```diff
@@ -58,24 +60,31 @@ async def unlock_restarted_vms(self, dataset_name):
                 if not path:
                     continue
 
-                if path.startswith(zvol_name_to_path(dataset_name) + '/') or path.startswith(f'/mnt/{dataset_name}/'):
+                if (
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/pool_/unlock.py:63]: @sonicaj can you please ensure that this is indeed the correct logic? Can VM be consuming a FILESYSTEM?

**Comment by themylogin** [comment]
time 3:00

File: src/middlewared/middlewared/plugins/pool_/unlock.py (hunk 1 above)
Comment (yocalebo) [reply src/middlewared/middlewared/plugins/pool_/unlock.py:63]: @Qubad786 too since he has done some work in this area

Comment (Qubad786) [reply src/middlewared/middlewared/plugins/pool_/unlock.py:63]: VM can yes consume a FILESYSTEM with `RAW` devices so this LGTM

File: tests/api2/test_pool_dataset_unlock_restart_vms.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        "inherit_encryption": False
+    }
+
+
+def test_restart_vm_on_dataset_unlock():
+    with dataset("test", encryption_props()) as ds:
+        call("pool.dataset.lock", ds, job=True)
+
+        with mock("vm.query", return_value=[
+            {"id": 1, "devices": [{"dtype": "DISK", "attributes": {"path": f"/mnt/{ds}/child"}}]},
```
[the remark this answers is not in the document]
Comment (themylogin) [reply tests/api2/test_pool_dataset_unlock_restart_vms.py:21]: Should I change `DISK` to `RAW` as you say here? https://github.com/truenas/middleware/pull/10408/files#r1070688935

Or you can just provide a list of possible options

[the remark this answers is not in the document]
Comment (Qubad786) [reply tests/api2/test_pool_dataset_unlock_restart_vms.py:21]: Yes either that or create a zvol and use it's path in DISK attribute

**Review by Qubad786 — APPROVED** [review]

**Review by yocalebo — APPROVED** [review]

**Pull request merged**